player/lua/select.lua

```diff
@@ -353,6 +353,99 @@ mp.add_key_binding(nil, "select-audio-device", function ()
     })
 end)
 
+local function get_history_entry(line, seen, history_file_path)
+    local entry = utils.parse_json(line)
+
+    if not entry then
+        mp.msg.warn(line .. " in " .. history_file_path .. " is not valid JSON.")
+        return
+    end
+
+    local time, path, title = unpack(entry)
+
+    if not path then
+        mp.msg.warn(line .. " in " .. history_file_path .. " has invalid data.")
```

Comment:
I would put this in else branch of `if entry then` in main loop. And instead printing `line`, which will leak history in log and also if it is malformed it may not be printable, I would log `i`, number of the line that fails.

Reply:
The else branch also runs if the entry is well formed but an already seen path.

Reply:
Reorganized it again so what you suggested here can be done and it also no longer keeps the file lines in memory.